### ISSUE TITLE
Question about normal estimation result (table 3)

### ISSUE THREAD

**Issue by Ryan-Zirui-Wang** [Yochengliu/Relation-Shape-CNN#25]
Hi @Yochengliu,

Many thanks for sharing the code. I have a question about the normal estimation result (table 3) in your paper. The table is attached below:

![image](https://user-images.githubusercontent.com/9391023/70568698-4f6cea80-1b90-11ea-8370-29b08b5f7646.png)

I am not sure how is the `error` defined? It looks like the mean cosine distance but I think the cosine distance should be -1 (i.e. the cosine similarity == 1) when the predictions are perfect? Did you apply an `abs()` to the cosine distance and compute the mean of all cosine distances?

Best,
Zirui

**Comment by Yochengliu** [comment]
Hi, I think the cosine distance = 1 - cosine similarity, that is, the cosine distance should be 0 when the predictions are perfect. This [link](https://en.wikipedia.org/wiki/Cosine_similarity) may help you, please have a check: ''The term cosine distance is often used for the complement in positive space, that is...''

**Comment by Ryan-Zirui-Wang** [comment]
Hi @Yochengliu, many thanks for your reply. Yeah `1 - cos_similarity` makes sense. Best, Zirui

**Comment by Ryan-Zirui-Wang** [comment]
Hi @Yochengliu,

Sorry to bother you again. A follow-up question about the surface normal estimation task: did you use the `cosine distance = 1 - cosine similarity` as loss term, or did you use the `abs(cosine distance)` as the loss term?

I'm asking this because when I read the pointnet paper, they said they use the `abs(cos_distance)`. I attached a screenshot of pointnet supplementary material below.

![image](https://user-images.githubusercontent.com/9391023/70916869-422e8080-2014-11ea-8b5c-fc007dbf1d12.png)

Best,
Zirui

**Comment by Yochengliu** [comment]
Hi, I use cosine distance = 1 - cosine similarity as loss, and personally, I think it may be unnecessary to apply the abs(), because the cosine distance >= 0.

**Comment by Ryan-Zirui-Wang** [comment]
Many thanks for your reply. Also, sorry, I didn't write the abs part correctly. I was thinking of having the abs in 
```
cos_distance = 1 - abs(cos_similarity)
```

In this case, flipped normals are not being punished. Anyway, I get your idea. Appreciate your fast response.

Best,
Zirui

**Comment by Ryan-Zirui-Wang** [comment]
Hi @Yochengliu,

Sorry to bother you again. I am having difficulties to reproduce the normal estimation results. The best mean cos distance I got is 0.65 so I think there must be some stupid mistakes I made. I wonder if there is any normal estimation code available when you train the RS-CNN? It doesn't even need to be runnable and I am just trying to read and check where I did wrong. Also, when you train the normal estimation network, did you use the ground truth normal in `modelnet40_ply_hdf5_2048.zip`?

Sorry for keep asking these small questions. Appreciate for your help.

Best,
Zirui
